### PR TITLE
Add nRF54L15 FLPR core support

### PR DESCRIPTION
**Pull request by jaz1-nordic**

### CODE REVIEW

File: soc/nordic/nrf54l/soc.c [hunk 1]
```diff
@@ -109,13 +113,15 @@ static int nordicsemi_nrf54l_init(void)
 		nrf_power_task_trigger(NRF_POWER, NRF_POWER_TASK_CONSTLAT);
 	}
 
+#if defined(NRF_APPLICATION)
```
Comment (masz-nordic) [inline soc/nordic/nrf54l/soc.c:116]: I suppose whole `nordicsemi_nrf54l_init` can be APP-specific, except for `SystemCoreClockUpdate`.

File: dts/arm/nordic/nrf54l15_cpuapp.dtsi [hunk 2]
```diff
@@ -4,88 +4,48 @@
...
-	};
+&rng {
+	status = "okay";
 };
 
-/* Disable by default to use GRTC */
-&systick {
-	status = "disabled";
+&cracen {
+	status = "okay";
```
Comment (masz-nordic) [inline dts/arm/nordic/nrf54l15_cpuapp.dtsi:50]: Are those nodes necessary?

File: soc/nordic/nrf54l/Kconfig.soc [hunk 3]
```diff
@@ -21,5 +21,14 @@ config SOC_NRF54L15_ENGA_CPUAPP
...
+	bool
+	select SOC_NRF54L15_ENGA
+	help
+	  NRF54L15 ENGA CPUFLPR
+
 config SOC
 	default "nrf54l15" if SOC_NRF54L15
+
+config HAS_HW_NRF_TEMP
+	bool
```
Comment (masz-nordic) [inline soc/nordic/nrf54l/Kconfig.soc:34]:
```suggestion
```

File: drivers/misc/nordic_vpr_launcher/nordic_vpr_launcher.c [hunk 4]
```diff
@@ -38,6 +41,9 @@ static int nordic_vpr_launcher_init(const struct device *dev)
 	}
 #endif
 
+#if CONFIG_SOC_NRF54L15_ENGA_CPUAPP
```
Comment (hubertmis) [inline drivers/misc/nordic_vpr_launcher/nordic_vpr_launcher.c:44]: nit: configuring SPU should not take place if building for NSPE:
```suggestion
#if defined(CONFIG_SOC_NRF54L15_ENGA_CPUAPP) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```

File: dts/arm/nordic/nrf54l15_cpuapp.dtsi [hunk 5]
```diff
@@ -4,88 +4,48 @@
...
+	ranges;
+};
 
-		peripheral@50000000 {
-			#address-cells = <1>;
-			#size-cells = <1>;
-			ranges = <0x0 0x50000000 0x10000000>;
+&grtc {
+	interrupts = <228 NRF_DEFAULT_IRQ_PRIORITY>,
+			<229 NRF_DEFAULT_IRQ_PRIORITY>; /* MPSL uses the GRTC_3_IRQn */
```
Comment (magp-nordic) [inline dts/arm/nordic/nrf54l15_cpuapp.dtsi:34]: This comment should probably be removed, there is no MPSL upstream.

Comment (hubertmis) [reply dts/arm/nordic/nrf54l15_cpuapp.dtsi:34]: Or maybe rephrased to sth like "reserved for Zero Latency IRQs"

File: dts/arm/nordic/nrf54l15_cpuapp.dtsi (hunk 2 above)
Comment (jaz1-nordic) [reply dts/arm/nordic/nrf54l15_cpuapp.dtsi]: I deleted them.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_USE_DT_CODE_PARTITION=y
+
+# Execute from SRAM
+CONFIG_XIP=n
+
+CONFIG_DEBUG=y
+
+CONFIG_NRF_GRTC_TIMER=y
```
Comment (anangl) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig:11]: This will be activated by the enabled `grtc` node.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+CONFIG_PINCTRL=y
```
Comment (anangl) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig:13]: This is redundant.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+# Enable GPIO
+CONFIG_GPIO=y
```
Comment (anangl) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig:16]: And this could be moved to the common `nrf54l15pdk_defconfig`.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15pdk/nrf54l15/cpuflpr
+name: nRF54L15-PDK-nRF54L15-Fast-Lightweight-Peripheral-Processor
+type: mcu
+arch: riscv
+toolchain:
+  - zephyr
+ram: 68
```
Comment (anangl) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.yaml:10]: Size of `cpuflpr_sram ` is 96 kB.

File: dts/arm/nordic/nrf54l15_cpuapp.dtsi [hunk 10]
```diff
@@ -4,88 +4,32 @@
...
-		hfxo: hfxo {
-			compatible = "nordic,nrf-hfxo";
-			#clock-cells = <0>;
-			clock-frequency = <DT_FREQ_M(32)>;
-		};
-	};
+/delete-node/ &cpuflpr;
+/delete-node/ &cpuflpr_rram;
+/delete-node/ &cpuflpr_sram;
+/delete-node/ &cpuflpr_clic;
```
[the remark this answers is not in the document]
Comment (jaz1-nordic) [reply dts/arm/nordic/nrf54l15_cpuapp.dtsi]: Line removed.

File: soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_enga_cpuflpr [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_NRF54L15_ENGA_CPUFLPR
+
+config RISCV_HAS_CPU_IDLE
+	bool
+
+config NUM_IRQS
+	int
+	default 480
```
[the remark this answers is not in the document]
Comment (jaz1-nordic) [reply soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_enga_cpuflpr]: I set this number to 286. It seems correct if we include 16 RISC interrupts.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_USE_DT_CODE_PARTITION=y
+
+# Execute from SRAM
+CONFIG_XIP=n
+
+CONFIG_DEBUG=y
```
[the remark this answers is not in the document]
Comment (jaz1-nordic) [reply boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig]: Removed.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig (hunk 8 above)
Comment (jaz1-nordic) [reply boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr_defconfig:12]: This file is missing after PR #71149.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+ #include "nrf54l15pdk_nrf54l15_cpuflpr_xip.dts"
+
+&cpuflpr_sram {
+	reg = <0x20028000 DT_SIZE_K(96)>;
+	ranges = <0x0 0x20028000 0x18000>;
+};
```
Comment (masz-nordic) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts:12]: This changes the RAM location, shouldn't the `cpuflpr_code_partition` be changed as well (from RRAM to SRAM?). Or am I missing something?

File: snippets/nordic-flpr-xip/README.rst [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+Nordic FLPR snippet with execution in place (nordic-flpr-xip)
+#############################################################
+
+Overview
+********
+
+This snippet allows users to build Zephyr with the capability to boot Nordic FLPR
+(Fast Lightweight Peripheral Processor) from application core.
+FLPR code is to be executed from RRAM, so the PPR image must be built
```
Comment (magp-nordic) [inline snippets/nordic-flpr-xip/README.rst:11]:
```suggestion
FLPR code is to be executed from RRAM, so the VPR image must be built
```
I guess i was supposed to be `VPR` or `FLPR`, not `PPR`.

**Comment by carlescufi** [comment]
@jaz1-nordic can you please rebase?

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts [hunk 15]
```diff
@@ -90,22 +42,57 @@
...
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		boot_partition: partition@0 {
+			label = "mcuboot";
+			reg = <0x0 DT_SIZE_K(64)>;
+		};
+		slot0_partition: partition@10000 {
+			label = "image-0";
+			reg = <0x10000 DT_SIZE_K(324)>;
```
Comment (pdunaj) [inline boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts:63]: why s and ns image have both 324k?

**Review by pdunaj — APPROVED** [review]

**Review by kl-cruz — APPROVED** [review]

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts (hunk 15 above)
Comment (jaz1-nordic) [reply boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts:63]: This is probably a question to @michalek-no ? #68393 
I did not change the partition layout. I only reduced them proportionally to create some space for the FLPR kernel.

File: boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts (hunk 13 above)
Comment (jaz1-nordic) [reply boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts]: It is not necessary. This is handled via the `XIP` Kconfig.

**Review by masz-nordic — APPROVED** [review]

**Review by magp-nordic — APPROVED** [review]

File: soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_enga_cpuflpr (hunk 11 above)
[the remark this answers is not in the document]
Comment (anangl) [reply soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_enga_cpuflpr]: Right. But I think you should then use 287, not 286, because the maximum IRQ index is 270 (CLOCK_POWER_IRQn).

File: dts/common/nordic/nrf54l15.dtsi [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+			};
+
+			uart00: uart@4a000 {
+				compatible = "nordic,nrf-uarte";
+				reg = <0x4a000 0x1000>;
+				interrupts = <74 NRF_DEFAULT_IRQ_PRIORITY>;
+				status = "disabled";
+			};
+
+			cpuflpr_vpr: vpr@4c000 {
```
Comment (anangl) [inline dts/common/nordic/nrf54l15.dtsi:171]: Please add entries for `cpuflpr_vpr` and `cpuflpr_clic` in `validate_base_addresses.c`.

**Review by anangl — DISMISSED** [review]

**Review by anangl — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
